src/braket/default_simulator/openqasm/circuit.py

```diff
@@ -0,0 +1,92 @@
+
+from braket.default_simulator.operation import GateOperation, KrausOperation
+from braket.default_simulator.operation_helpers import from_braket_instruction
+from braket.default_simulator.result_types import _from_braket_observable
+
+
+class Circuit:
+    """
+    This is a lightweight analog to braket.ir.jaqcd.program_v1.Program.
+    The Interpreter compiles to an IR to hand off to the simulator,
+    braket.default_simulator.state_vector_simulator.StateVectorSimulator, for example.
+    Our simulator module takes in a circuit specification that satisfies the interface
+    implemented by this class.
+    """
+
+    def __init__(
+        self,
+        instructions: Optional[List[GateOperation]] = None,
+        results: Optional[List[Results]] = None,
+    ):
+        self.instructions = []
+        self.results = []
+        self.qubit_set = set()
+
+        if instructions:
+            for instruction in instructions:
+                self.add_instruction(instruction)
+
+        if results:
+            for result in results:
+                self.add_result(result)
+
+    def add_instruction(self, instruction: [GateOperation, KrausOperation]) -> None:
+        """
+        Add instruction to the circuit.
+
+        Args:
+            instruction (GateOperation): Instruction to add.
+        """
+        self.instructions.append(instruction)
+        self.qubit_set |= set(instruction.targets)
+
+    def add_result(self, result: Results) -> None:
+        """
+        Add result type to the circuit.
+
+        Args:
+            result (Results): Result type to add.
+        """
+        self.results.append(result)
+        if isinstance(result, OptionalMultiTarget) and result.targets is not None:
+            self.qubit_set |= set(result.targets)
+
+    @property
+    def num_qubits(self) -> int:
+        return len(self.qubit_set)
+
+    @property
+    def basis_rotation_instructions(self):
+        """
+        This function assumes all observables are commuting.
+        """
+
+        basis_rotation_instructions = []
+        measured_qubits = set()
+
+        for result in self.results:
+            if isinstance(result, Observable):
+                observables = result.observable
+                actual_targets = result.targets or range(self.num_qubits)
+
+                if set(actual_targets).issubset(measured_qubits):
+                    continue
+                elif set(actual_targets).isdisjoint(measured_qubits):
+                    braket_obs = _from_braket_observable(observables, result.targets)
+                    diagonalizing_gates = braket_obs.diagonalizing_gates(self.num_qubits)
+                    basis_rotation_instructions.extend(diagonalizing_gates)
+                    measured_qubits |= set(actual_targets)
+                else:
+                    # this shouldn't impact any circuits coming from the BDK
+                    raise NotImplementedError("Partially measured observable target")
```

Comment:
Nit: I'm just being annoying, I prefer not to use `continue` if possible. Take it or leave it.
```suggestion
                if set(actual_targets).isdisjoint(measured_qubits):
                    braket_obs = _from_braket_observable(observables, result.targets)
                    diagonalizing_gates = braket_obs.diagonalizing_gates(self.num_qubits)
                    basis_rotation_instructions.extend(diagonalizing_gates)
                    measured_qubits |= set(actual_targets)
                elif not set(actual_targets).issubset(measured_qubits):
                    # this shouldn't impact any circuits coming from the BDK
                    raise NotImplementedError("Partially measured observable target")
```